src/base/utils/tar.cpp

```diff
@@ -0,0 +1,230 @@
+/*
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2020  sledgehammer999 <hammered999@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "tar.h"
+
+#include <QByteArray>
+#include <QCoreApplication>
+#include <QHash>
+#include <QString>
+
+#include "base/logger.h"
+#include "base/utils/bytearray.h"
+#include "base/utils/fs.h"
+
+namespace
+{
+    const int TAR_HEADER_SIZE = 512;
+
+    const int TAR_FILENAME_OFFSET = 0;
+    const int TAR_FILESIZE_OFFSET = 124;
+    const int TAR_CHECKSUM_OFFSET = 148;
+    const int TAR_FILETYPE_OFFSET = 156;
+    const int TAR_FILENAME_PREFIX_OFFSET = 345;
+
+    const int TAR_FILENAME_SIZE = 100;
+    const int TAR_FILESIZE_SIZE = 12;
+    const int TAR_CHECKSUM_SIZE = 8;
+    const int TAR_FILENAME_PREFIX_SIZE = 155;
+
+    class TarEntry
+    {
+    public:
+
+        enum Types : unsigned char
+        {
+            REG_FILE_NULL = '\0',
```

Comment:
Other functions require passing the `'\0'` literal because of the compiler. It doesn't know which overload to call (eg for startsWith()).
Furthermore, in the spec these are treated as ASCII values, so for keeping things uniform, I will leave it as ASCII NUL.